featurebyte/session/bigquery.py

```diff
@@ -0,0 +1,707 @@
+            next_page_token = table_iterator.next_page_token
+            if not next_page_token:
+                break
+        return output
+
+    @staticmethod
+    def _convert_to_internal_variable_type(  # pylint: disable=too-many-return-statements
+        bigquery_typecode: str, scale: int = 0, mode: str | None = None
+    ) -> DBVarType:
+        """
+        Convert BigQuery data type to internal variable type
+
+        Parameters
+        ----------
+        bigquery_typecode: str
+            BigQuery data type
+        scale: int
+            Scale
+        mode: str | None
+            Mode
+
+        Returns
+        -------
+        DBVarType
+        """
+        if bigquery_typecode in {SqlTypeNames.NUMERIC, SqlTypeNames.BIGNUMERIC}:
+            if scale > 0:
+                return DBVarType.FLOAT
+            return DBVarType.INT
+        db_vartype = db_vartype_mapping.get(bigquery_typecode, DBVarType.UNKNOWN)
+        if mode == "REPEATED":
+            db_vartype = DBVarType.ARRAY
+        elif db_vartype == DBVarType.UNKNOWN:
+            logger.warning(f"BigQuery: Not supported data type '{bigquery_typecode}'")
+        return db_vartype
+
+    @staticmethod
+    def _get_pyarrow_type(
+        datatype: str, precision: int = 0, scale: int = 0, mode: str | None = None
+    ) -> pa.DataType:
+        """
+        Get pyarrow type from BigQuery data type
+
+        Parameters
+        ----------
+        datatype: str
+            BigQuery data type
+        precision: int
+            Precision
+        scale: int
+            Scale
+        mode: str | None
+            Mode
+
+        Returns
+        -------
+        pa.DataType
+        """
+        if datatype == SqlTypeNames.INTERVAL:
+            pyarrow_type = pa.int64()
+        elif datatype in {SqlTypeNames.NUMERIC, SqlTypeNames.BIGNUMERIC}:
+            if scale > 0:
+                pyarrow_type = pa.decimal128(precision, scale)
+            else:
+                pyarrow_type = pa.int64()
+        elif mode == "REPEATED":
+            pyarrow_type = pa.string()
+        else:
+            pyarrow_type = pa_type_mapping.get(datatype)
+
+        if not pyarrow_type:
+            # warn and fallback to string for unrecognized types
+            logger.warning("Cannot infer pyarrow type", extra={"datatype": datatype})
+            pyarrow_type = pa.string()
+        return pyarrow_type
+
+    def _get_schema_from_cursor(self, cursor: Any) -> pa.Schema:
+        """
+        Get schema from a cursor
+
+        Parameters
+        ----------
+        cursor: Any
+            Cursor to fetch data from
+
+        Returns
+        -------
+        pa.Schema
+        """
+        # get schema from query job as cursor description is missing REPEATED information
+        try:
+            schema = self._client._get_query_results(cursor.query_job.job_id, DEFAULT_RETRY).schema  # pylint: disable=protected-access
```

Comment:
Does `_get_query_results` retrieve rows as well? Will it be an issue if the query result is large?

Reply:
Indeed I will set `page_size` to 1